firedrake/variational_solver.py

```diff
@@ -258,22 +253,36 @@ def solve(self, bounds=None):
         self._ctx.set_function(self.snes)
         self._ctx.set_jacobian(self.snes)
 
-        # Make sure appcontext is attached to the DM before we solve.
-        dm = self.snes.getDM()
+        # Make sure appcontext is attached to every coefficient DM before we solve.
+        problem = self._problem
+        problem_dms = [self.snes.getDM()]
+        seen = {problem.u.function_space()}
+        coefficients = problem.F.coefficients() + problem.J.coefficients()
+        if problem.Jp is not None:
+            coefficients = coefficients + problem.Jp.coefficients()
+        for val in chain(coefficients, (bc.function_arg for bc in problem.bcs)):
+            if isinstance(val, (function.Function, cofunction.Cofunction)):
+                V = val.function_space()
+                if V not in seen:
+                    problem_dms.append(V.dm)
+                    seen.add(V)
+
         for dbc in self._problem.dirichlet_bcs():
             dbc.apply(self._problem.u)
 
         if bounds is not None:
             lower, upper = bounds
             with lower.dat.vec_ro as lb, upper.dat.vec_ro as ub:
                 self.snes.setVariableBounds(lb, ub)
+
         work = self._work
-        with self._problem.u.dat.vec as u:
+        with problem.u.dat.vec as u:
             u.copy(work)
             with ExitStack() as stack:
                 # Ensure options database has full set of options (so monitors
                 # work right)
-                for ctx in chain((self.inserted_options(), dmhooks.add_hooks(dm, self, appctx=self._ctx)),
+                for ctx in chain((self.inserted_options(),
+                                  *(dmhooks.add_hooks(dm, self, appctx=self._ctx) for dm in reversed(problem_dms))),
```

Comment:
@wence- Apparently the order of the contexts matters, we need to visit the solution DM after all other coefficients. Does this look good?

Reply:
:shrug: Probably? Hooks are torn down in LIFO order, but I don't know why that is important in this scenario. What happens if you don't do it this way round?

Reply:
Minor, no need to splat the generator since you're in a `chain`, so:

```
chain([self.inserted_options()], (dmhooks.add_hooks(...) for dm in ...), self._transfer_operators)
```
I suppose.

Reply:
I can't put in a generator, either I pass options and hooks together in the same tuple as I had before, or I use your suggestion with `tuple` in front of the generator.

Reply:
> 🤷 Probably? Hooks are torn down in LIFO order, but I don't know why that is important in this scenario. What happens if you don't do it this way round?

If I don't use `reverse` then the appctx does not get attached properly and I get the warning "Creating new TransferManager".

Reply:
Huh, odd, anyway, I think that is fine.